crates/ruma-events/src/room/join_rules.rs

```diff
@@ -25,67 +23,130 @@ pub type JoinRulesEvent = StateEvent<JoinRulesEventContent>;
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(tag = "join_rule")]
 pub enum JoinRule {
     /// A user who wishes to join the room must first receive an invite to the room from someone
     /// already inside of the room.
+    #[serde(rename = "invite")]
     Invite,
 
     /// Reserved but not yet implemented by the Matrix specification.
+    #[serde(rename = "knock")]
     Knock,
 
     /// Reserved but not yet implemented by the Matrix specification.
+    #[serde(rename = "private")]
     Private,
 
     /// Users can join the room if they are invited, or if they meet any of the conditions
     /// described in a set of [`AllowRule`]s.
     #[cfg(feature = "unstable-pre-spec")]
-    Restricted,
+    #[serde(rename = "restricted")]
+    Restricted(Restricted),
 
     /// Anyone can join the room without any prior action.
+    #[serde(rename = "public")]
     Public,
 
     #[doc(hidden)]
-    _Custom(String),
+    _Custom(CustomJoinRule),
 }
 
 impl JoinRule {
     /// Creates a string slice from this `JoinRule`.
     pub fn as_str(&self) -> &str {
-        self.as_ref()
+        match self {
+            JoinRule::Invite => "invite",
+            JoinRule::Knock => "knock",
+            JoinRule::Private => "private",
+            #[cfg(feature = "unstable-pre-spec")]
+            JoinRule::Restricted(_) => "restricted",
+            JoinRule::Public => "public",
+            JoinRule::_Custom(rule) => &rule.join_rule,
+        }
+    }
+}
+
+#[doc(hidden)]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct CustomJoinRule {
+    join_rule: String,
+    #[serde(flatten)]
+    extra: BTreeMap<String, JsonValue>,
+}
+
+impl<'de> Deserialize<'de> for JoinRule {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        fn from_raw_json_value<T: DeserializeOwned, E: Error>(raw: &RawJsonValue) -> Result<T, E> {
+            serde_json::from_str(raw.get()).map_err(E::custom)
+        }
+
+        let json: Box<RawJsonValue> = Box::deserialize(deserializer)?;
+
+        #[derive(Deserialize)]
+        struct ExtractType<'a> {
+            join_rule: Option<Cow<'a, str>>,
+        }
+
+        let join_rule = serde_json::from_str::<ExtractType>(json.get())
+            .map_err(serde::de::Error::custom)?
+            .join_rule;
+
+        match join_rule.as_deref() {
+            Some("invite") => Ok(JoinRule::Invite),
+            Some("knock") => Ok(JoinRule::Knock),
+            Some("private") => Ok(JoinRule::Private),
+            #[cfg(feature = "unstable-pre-spec")]
+            Some("restricted") => from_raw_json_value(&json).map(Self::Restricted),
+            Some("public") => Ok(JoinRule::Public),
+            Some(_) => from_raw_json_value(&json).map(Self::_Custom),
+            None => Err(D::Error::missing_field("join_rule")),
+        }
```

Comment:
I think this is slightly easier to read with the error handling moved up:

```suggestion
        let join_rule = serde_json::from_str::<ExtractType>(json.get())
            .map_err(serde::de::Error::custom)?
            .join_rule
            .ok_or_else(|| D::Error::missing_field("join_rule"))?;

        match join_rule.as_str() {
            "invite" => Ok(JoinRule::Invite),
            "knock" => Ok(JoinRule::Knock),
            "private" => Ok(JoinRule::Private),
            #[cfg(feature = "unstable-pre-spec")]
            "restricted" => from_raw_json_value(&json).map(Self::Restricted),
            "public" => Ok(JoinRule::Public),
            _ => Ok(Self::_Custom(join_rule)),
        }
```

Also I didn't change this above, but please use `Self::` consistently (instead of some `Self::`s and some `JoinRule::`s).